tutorials/basic-topics/sorting/merge-sort.md

```diff
@@ -11,4 +11,504 @@ keywords:
+                int end = Math.min(i + 2 * size - 1, nums.length - 1);
+                merge(nums, i, mid, end);
+            }
+        }
+    }
+
+    private void merge(int[] nums, int l, int mid, int r) {
+        int[] tmp = new int[r - l + 1];
+        int i = l, j = mid + 1, k = 0;
+        while (i <= mid || j <= r) {
+            if (i > mid || j <= r && nums[i] > nums[j]) {
+                tmp[k++] = nums[j++];
+            } else {
+                tmp[k++] = nums[i++];
+            }
+        }
+        System.arraycopy(tmp, 0, nums, l, r - l + 1);
+    }
+}
+```
+</TabItem>
+</Tabs>
+
+**Time Complexity** : Best & Worst & Average is   $O(n \log_{2} n)$
+
+**Space Complexity** :  $O(n)$ 
+
+**Stablity**       :  Stable
+
+**Is In Place**      :  Not in place
+
+**When  not to use**: When the array is too large
+
+
+Merging of n elements takes n time and since each time the array is cut into half it takes $\log_{2}n$ time to reach the top so **total time complexity is** $O(n\log_{2} n)$.
+
+
+![image](https://cdn.kastatic.org/ka-perseus-images/5fcbebf66560d8fc490de2a0d8a0e5b1d65c5c54.png)
+
+Image taken from [Khan Academy](https://cdn.kastatic.org/)
+
+
+Note: **Mergesort is also useful for sorting linked lists**
+
+## [**Leetcode Problem 148_Sort List**](https://leetcode.com/problems/sort-list/)
+
+**Problem Statement** 
+
+Given the head of a linked list, return the list after sorting it in ascending order.
+
+**Test Cases**
+ 
+Test Case 1
+```
+Input: head = [4,2,1,3]
+Output: [1,2,3,4]
+```
+![image](https://assets.leetcode.com/uploads/2020/09/14/sort_list_1.jpg)
+ 
+Test Case 2
+
+```
+Input: head = [-1,5,3,4,0]
+Output: [-1,0,3,4,5]
+```
+
+Here we can follow both **top-down and bottom-up merge sort**. I have already discussed the algorithm for arrays in a detailed way.
+
+### Intuition for the problem
+```
+----Example of sorting of Linked List --------
+
+3->8->12->null
+1->10->23->null
+
+After merging we get :
+
+1->3->8->10->12->23->null.
+
+------ Hence Sorted---------------------------
+
+1. The two lists to be merged must be ordered respectively.
+2. We can only start to merge two lists that only have one element.
+3. Then we get an ordered list that has two elements
+4. Do this again (that is "recursion").
+
+ Bottom-Up Approach 
+
+    -1-> 5->  3-> 4->   0-> null
+      \  /       \ /         /
+   -1->5->null 0->3->4->null 
+      \        /         
+    -1-> 0->3-> 4-> 5->null   (sorted) 
+
+----- Finally Sorted ---------------------------------
+
+ Top Down Approach
+
+   -1-> 5->  3-> 4-> 0-> null
+      /          \ 
+-1-> 5->null      3-> 4->0->null
+ /       \           /      \  
+-1->null 5->null 3->null   4->0->null
+  \   /        \          /     \
+ -1-> 5->null  3->null   4->null 0->null
+   \             \        \   /
+    \           3->null     0-> 4->null
+     \              \      /  
+      \           0->3->4->null
+       \          /
+    -1->0->3->4->5->null    (sorted)
+```
```

Comment:
the presentation of this part is not that nice. Especially the linked list part. Suggest to make some diagrams.

Reply:
@wingkwong I am resolving this part by removing it! Since drawing diagrams needs to be done by Mermaid, I guess GitHub supports it, but it's too time-consuming to draw a tree with a linked list.

Reply:
What is Mermaid? 

You can use lucidchart or draw.io to draw the diagram and upload as images and paste the link there.